spec/app/ics-100-atomic-swap/README.md

```diff
@@ -0,0 +1,568 @@
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyPortIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  counterpartyVersion: string) => (version: string, err: Error) {
+  // only unordered channels allowed
+  abortTransactionUnless(order === UNORDERED)
+  // assert that version is "ics100-1"
+  abortTransactionUnless(counterpartyVersion === "ics100-1")
+
+  return "ics100-1", nil
+}
+```
+
+```typescript
+function onChanOpenAck(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  counterpartyVersion: string
+) {
+  // port has already been validated
+  // assert that counterparty selected version is "ics100-1"
+  abortTransactionUnless(counterpartyVersion === "ics100-1");
+}
+```
+
+```typescript
+function onChanOpenConfirm(portIdentifier: Identifier, channelIdentifier: Identifier) {
+  // accept channel confirmations, port has already been validated, version has already been validated
+}
+```
+
+```typescript
+function onChanCloseInit(portIdentifier: Identifier, channelIdentifier: Identifier) {
+  // always abort transaction
+  abortTransactionUnless(FALSE);
+}
+```
+
+```typescript
+function onChanCloseConfirm(portIdentifier: Identifier, channelIdentifier: Identifier) {
+  // no action necessary
+}
+```
+
+#### Packet relay
+
+`sendAtomicSwapPacket` must be called by a transaction handler in the module which performs appropriate signature checks, specific to the account owner on the host state machine.
+
+```typescript
+function sendAtomicSwapPacket(
+    swapPacket AtomicSwapPacketData, 
+    sourcePort, sourceChannel, timeoutHeight, timeoutTimestamp: int64
+) {
+    // send packet using the interface defined in ICS4
+    handler.sendPacket(
+      getCapability("port"),
+      sourcePort,
+      sourceChannel,
+      timeoutHeight,
+      timeoutTimestamp,
+      swapPacket.getBytes(), // Should be proto marshalled bytes.
+    )
+}
+```
+
+`onRecvPacket` is called by the routing module when a packet addressed to this module has been received.
+
+```typescript
+function onRecvPacket(packet channeltypes.Packet) {
+  switch packet.type {
+      case TYPE_MSG_MAKE_SWAP:
+        const make_msg = protobuf.decode(packet.bytes)
+        
+        // check if buy_token is native token on the taker chain
+        const supply = bank_keeper.getSuppy(make_msg.buy_token.denom)
+        abortTransactionUnless(supply > 0)
+        
+        // create and save order on the taker chain.
+        const order = OrderBook.createOrder(msg)
+        order.status = Status.SYNC
+        order.port_id_on_taker_chain = packet.destinationPort
+        order.channel_id_on_taker_chain = packet.destinationChannel
+        //saves order to store
+        store.save(order)
+        break;
+      case TYPE_MSG_TAKE_SWAP:
+        const take_msg = protobuf.decode(packet.bytes)
+        const order = OrderBook.findOrderById(take_msg.order_id)
```

Comment:
> For example, that the order ID used in the maker is different from the order ID in the taker.

Because the order id is the hash of the data packet